src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -246,7 +246,7 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
     case AndType(l, r) =>
       simplify(l, theMap) & simplify(r, theMap)
     case OrType(l, r) =>
```

Comment:
Simplified is called in lots of places for types that are not written. So it should not to the lub with keepSingletons = true.